dask/dataframe/io/parquet/core.py

```diff
@@ -701,60 +759,56 @@ def to_parquet(
         **kwargs_pass,
     )
 
-    # check name_function is valid
-    if name_function is not None and not callable(name_function):
-        raise ValueError("``name_function`` must be a callable with one argument.")
-
-    # Use i_offset and df.npartitions to define file-name list
-    filenames = [
-        f"part.{i + i_offset}.parquet"
-        if name_function is None
-        else name_function(i + i_offset)
-        for i in range(df.npartitions)
-    ]
-
-    if name_function is not None and len(set(filenames)) < len(filenames):
-        raise ValueError("``name_function`` must produce unique filenames.")
+    # Check that custom name_function is valid,
+    # and that it will produce unique names
+    if name_function is not None:
+        if not callable(name_function):
+            raise ValueError("``name_function`` must be a callable with one argument.")
+        filenames = [name_function(i + i_offset) for i in range(df.npartitions)]
+        if len(set(filenames)) < len(filenames):
+            raise ValueError("``name_function`` must produce unique filenames.")
 
-    # Construct IO graph
-    dsk = {}
-    name = "to-parquet-" + tokenize(
-        df,
-        fs,
-        path,
-        append,
-        ignore_divisions,
-        partition_on,
-        division_info,
-        index_cols,
-        schema,
-    )
-    part_tasks = []
+    # Create Blockwise layer for parquet-data write
     kwargs_pass["fmd"] = meta
     kwargs_pass["compression"] = compression
     kwargs_pass["index_cols"] = index_cols
     kwargs_pass["schema"] = schema
-    for d, filename in enumerate(filenames):
-        dsk[(name, d)] = (
-            apply,
-            engine.write_partition,
-            [
-                (df._name, d),
-                path,
-                fs,
-                filename,
-                partition_on,
-                write_metadata_file,
-            ],
-            toolz.merge(kwargs_pass, {"head": True}) if d == 0 else kwargs_pass,
-        )
-        part_tasks.append((name, d))
-
-    final_name = "metadata-" + name
-    # Collect metadata and write _metadata
+    data_write = df.map_partitions(
+        ToParquetFunctionWrapper(
+            engine,
+            path,
+            fs,
+            partition_on,
+            write_metadata_file,
+            i_offset,
+            name_function,
+            kwargs_pass,
+        ),
+        token="to-parquet-"
+        + tokenize(
```

Comment:
What if you implement `__dask_tokenize__` on `ToParquetFunctionWrapper`? Then I'd imagine the tokenization logic already in `map_partitions` would just work.

Reply:
If we don't pass in `token=` explicitly, then `map_partitions` will produce a layer name of the form: `f"{funcname(func)}-{tokenize(func, meta, *args, **kwargs)}"`.  Therefore, we would also need to modify the `ToParquetFunctionWrapper` name to be `"to-parquet"`.  Is it worthwhile to add these defintions to `ToParquetFunctionWrapper` when we can just establish the name here?

Reply:
Wait, the `token=` argument to `map_partitions` is very much mis-named. It should really be called `name`:https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/dataframe/core.py#L6028-L6037

With what you have here, you'll get two tokens on the final name (`map_partitions` will append one automatically). I think just `token="to-parquet"` and a `__dask_tokenize__` method on `ToParquetFunctionWrapper` (and `BlockwiseDep`) is what you want here.

Reply:
>Wait, the token= argument to map_partitions is very much mis-named. It should really be called name

Agree

>With what you have here, you'll get two tokens on the final name

Agree - I didn't realize this before, and I certainly don't like this.

>I think just token="to-parquet" and a __dask_tokenize__ method on ToParquetFunctionWrapper (and BlockwiseDep) is what you want here.

If we specify `token="to-parquet"`, then I don't think `ToParquetFunctionWrapper.__dask_tokenize__` will ever be used. In order to avoid making changes in `map_partitions`, I vote that we just define the `__repr__` of `ToParquetFunctionWrapper` to be `"to-parquet"`, and define `__dask_tokenize__` (as you suggested).

Reply:
> If we specify `token="to-parquet"`, then I don't think `ToParquetFunctionWrapper.__dask_tokenize__` will ever be used.

Don't think this is true: `ToParquetFunctionWrapper` will be part of `args` here, so it'll get tokenized as usual.

https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/dataframe/core.py#L6032-L6033